go/controller/api/v1alpha1/federatedcredential_types.go

```diff
@@ -0,0 +1,86 @@
+package v1alpha1
+
+import (
+	"github.com/samber/lo"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	console "github.com/pluralsh/console/go/client"
+)
+
+func init() {
+	SchemeBuilder.Register(&FederatedCredentialList{}, &FederatedCredential{})
+}
+
+// +kubebuilder:object:root=true
+
+// FederatedCredentialList contains a list of FederatedCredential
+type FederatedCredentialList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []FederatedCredential `json:"items"`
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Cluster
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="ID",type="string",JSONPath=".status.id",description="ID of the FederatedCredential in the Console API."
+
+// FederatedCredential is a way to authenticate users from an external identity provider.
+type FederatedCredential struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   FederatedCredentialSpec `json:"spec,omitempty"`
+	Status Status                  `json:"status,omitempty"`
+}
+
+func (in *FederatedCredential) Diff(hasher Hasher) (changed bool, sha string, err error) {
+	currentSha, err := hasher(in.Spec)
+	if err != nil {
+		return false, "", err
+	}
+
+	return !in.Status.IsSHAEqual(currentSha), currentSha, nil
+}
+
+func (in *FederatedCredential) Attributes() console.FederatedCredentialAttributes {
+	claimsLike := in.Spec.ClaimsLike
+	if claimsLike == nil || len(*claimsLike) == 0 {
+		// If claimsLike is not set, we default to an empty JSON object.
+		// This is to ensure that the API can handle it gracefully.
+		claimsLike = lo.ToPtr("{}")
+	}
+
+	return console.FederatedCredentialAttributes{
+		Issuer:     in.Spec.Issuer,
+		Scopes:     lo.ToSlicePtr(in.Spec.Scopes),
+		ClaimsLike: claimsLike,
+	}
+}
+
+func (in *FederatedCredential) SetCondition(condition metav1.Condition) {
+	meta.SetStatusCondition(&in.Status.Conditions, condition)
+}
+
+// FederatedCredentialSpec defines the desired state of FederatedCredential.
+type FederatedCredentialSpec struct {
+	// Issuer is the URL of the identity provider that issues the tokens.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Issuer is immutable"
+	// +kubebuilder:validation:Required
+	Issuer string `json:"issuer"`
+
+	// Scopes are the scopes that the credential will request from the identity provider.
+	// +kubebuilder:validation:Optional
+	Scopes []string `json:"scopes,omitempty"`
+
+	// ClaimsLike is a JSON expression that matches the claims in the token.
+	// TODO: describe the syntax of this expression.
+	// +kubebuilder:validation:Optional
+	ClaimsLike *string `json:"claimsLike,omitempty"`
```

Comment:
We should make this a raw extension object so users don't have to manually json format (then you json encode it w/in the operator)

Reply:
makes sense, done